contracts/epochs/FlowDKG.cdc

```diff
@@ -0,0 +1,274 @@
+    // Each resource instance is good for all future potential epochs, but will
+    // only be valid if the node operator has been confirmed as a consensus node for the next epoch.
+    pub resource Participant {
+
+        pub let nodeID: String
+
+        // Posts a whiteboard message to the contract
+        pub fun postMessage(_ content: String) {
+            pre {
+                FlowDKG.participantIsRegistered(self.nodeID): "Cannot send whiteboard message if not registered for the current epoch"
+            }
+
+            // create the message struct
+            let message = Message(nodeID: self.nodeID, content: content)
+
+            // add the message to the message record
+            FlowDKG.whiteboardMessages.append(message)
+
+            emit BroadcastMessage(nodeID: self.nodeID, content: content)
+
+        }
+
+        /// Sends the final key vector submission. 
+        /// Can only be called by consensus nodes that are registered.
+        pub fun sendFinalSubmission(_ submission: [String]) {
+            pre {
+                FlowDKG.participantIsRegistered(self.nodeID): "Cannot send final submission if not registered for the current epoch"
+                FlowDKG.nodeHasSubmitted(self.nodeID)!.length == 0: "Cannot submit a final submission twice"
+                submission.length == FlowDKG.nodeFinalSubmission.keys.length + 1: "Submission must have number of elements equal to the number of nodes participating in the DKG"
+            }
+
+            var finalSubmissionIndex = 0
+
+            // iterate through all the existing unique submissions
+            // If this participant's submission matches one of the existing ones,
+            // add to the counter for that submission
+            // Otherwise, track the new submission and set its counter to 1
+            for existingSubmission in FlowDKG.uniqueFinalSubmissions {
+
+                // If the submissions are equal,
+                // update the counter for this submission and emit the event
+                if FlowDKG.submissionsEqual(existingSubmission, submission) {
+                    FlowDKG.uniqueFinalSubmissionCount[finalSubmissionIndex] = FlowDKG.uniqueFinalSubmissionCount[finalSubmissionIndex]! + 1 as UInt64
+                    emit FinalSubmission(nodeID: self.nodeID, submission: submission)
+                    break
+                }
+
+                // update the index counter
+                finalSubmissionIndex = finalSubmissionIndex + 1
+
+                // If no matches were found, add this submission as a new unique one
+                // and emit an event
+                if finalSubmissionIndex == FlowDKG.uniqueFinalSubmissions.length {
+                    FlowDKG.uniqueFinalSubmissionCount[finalSubmissionIndex] = 0
+                    FlowDKG.uniqueFinalSubmissions.append(submission)
+                    emit FinalSubmission(nodeID: self.nodeID, submission: submission)
```

Comment:
ok. I usually default to adding events anywhere where I think it might be useful to do it, but if this isn't necessary, I'll remove it

Reply:
Some extra context @tarakby: this is a regular event (not a service event) and does not bubble up to the protocol software / network. It's just accessible for anyone who wants to query it.

Reply:
It's good that is doesn't notify the protocol/network. But since the info won't be useful for any party, it's maybe worth it to save the little overhead.